File: src/main/java/com/somefrills/features/core/Features.java
package com.somefrills.features.core;

public final class Features {

    private static final AbstractFeature[] INSTANCES = FeaturesRegistry.INSTANCES;
    private static final ClassValue<AbstractFeature> CLASS_TO_INSTANCE = FeaturesRegistry.CLASS_TO_INSTANCE;

    private static volatile boolean initialized = false;

    public static boolean isInitialized() {
        return initialized;
    }

    public static void init() {
        FeaturesRegistry.init();
        for (AbstractFeature feature : INSTANCES) {
            feature.initialize();
        }
        initialized = true;
    }

    public static <T extends AbstractFeature> T get(Class<T> clazz) {
        AbstractFeature feature = CLASS_TO_INSTANCE.get(clazz);
        if (feature == null) return null;

        @SuppressWarnings("unchecked")
        T instance = (T) feature;

        return instance;
    }

    public static <T extends AbstractFeature> boolean isActive(Class<T> clazz) {
        var instance = get(clazz);
        if (instance == null) return false;
        return instance.isActive();
    }
}

File: src/main/kotlin/com/somefrills/features/core/Features.kt
package com.somefrills.features.core

object Features {

    private val registry = FeaturesRegistry
    private var initialized = false

    @JvmStatic
    fun isInitialized(): Boolean {
        return initialized
    }

    @JvmStatic
    fun init() {
        registry.init()

        for (f in registry.INSTANCES.filterNotNull()) {
            f.initialize()
        }

        initialized = true
    }

    @JvmStatic
    @Suppress("UNCHECKED_CAST")
    fun <T : AbstractFeature> get(clazz: Class<T>): T {
        val f = registry.CLASS_TO_INSTANCE.get(clazz)
            ?: error("No feature for $clazz")
        return f as T
    }

    @JvmStatic
    fun <T : AbstractFeature> isActive(clazz: Class<T>): Boolean {
        return get(clazz).isActive()
    }

    @JvmStatic
    fun all(): List<AbstractFeature> {
        return registry.INSTANCES.filterNotNull()
    }
}